src/AutoRest.CSharp/Common/Output/Models/Requests/RestClientMethod.cs

```diff
@@ -16,51 +16,83 @@ public RestClientMethod(string name, string? description, CSharpType? returnType
         {
             Name = name;
             Request = request;
-            Parameters = parameters;
             Responses = responses;
             Description = description;
             ReturnType = returnType;
             HeaderModel = headerModel;
             BufferResponse = bufferResponse;
             Accessibility = accessibility;
             Operation = operation;
+
+            PathParameters = getPathParameters(parameters);
+            NonPathParameters = getNonPathParameters(parameters);
+
+            Parameters = SortParameters(parameters, new Queue<Parameter>(PathParameters));
         }
 
         public string Name { get; }
         public string? Description { get; }
         public Request Request { get; }
-        public Parameter[] Parameters { get; }
+        public List<Parameter> Parameters { get; }
         public Response[] Responses { get; }
         public DataPlaneResponseHeaderGroupType? HeaderModel { get; }
         public bool BufferResponse { get; }
         public CSharpType? ReturnType { get; }
         public string Accessibility { get; }
         public Operation Operation { get; }
 
-        public List<Parameter> NonPathParameters
+        public List<Parameter> PathParameters { get; }
+
+        public List<Parameter> NonPathParameters { get; }
+
+        private List<Parameter> getPathParameters(Parameter[] parameters)
         {
-            get
-            {
-                var pathSegments = this.Request.PathParameterSegments;
+            List<Parameter> pathParameters = new();
+            // Path segments are ordered from left to right, so iterate segments first will keep the sequence of
+            // path parameters in URL path
+            Request.PathParameterSegments.ForEach(pathSegment => {
+                var parameter = parameters.FirstOrDefault(p => pathSegment.Value.Reference.Type.Name == p.Type.Name &&
+                    pathSegment.Value.Reference.Name == p.Name);
+                if (parameter != null)
+                {
+                    pathParameters.Add(parameter);
+                }
+            });
 
-                return this.Parameters.Where(parameter => !pathSegments.Any(
-                    pathSegment => pathSegment.Value.Reference.Type.Name == parameter.Type.Name &&
-                    pathSegment.Value.Reference.Name == parameter.Name)
-                ).ToList();
-            }
+            return pathParameters;
         }
```

Comment:
The only drawback is that the algorithm complexity is `O(n^2logn)`, since getting `key` for each item in `OrderBy` is an `O(n)` operation.
But considering the small data volume, it's still worthy.

Reply:
The problem with this sorting algorithm is that we will move non-path parameters to the bottom. So the original `parameters` array is split into 2 parts:
- `PathParameters`: from the beginning, and sorted
- `NonPathParameters`: the remaining, and preserve the original order

That will work if it's guaranteed that `PathParameters` is always from the beginning.

Here I choose a more conservative strategy: only sort `PathParameters` **in their current indices** in `parameters`. Let's assume we have 3 path parameters: P1, P2, P3, and 3 non-path parameters: A, B, C.
If the original `parameters` array is `A, P3, P2, B, C, P1`. What I want to achieve is that finally we get `A, P1, P2, B, C, P3`. The suggested algorithm will get `P1, P2, P3, A, B, C`.

So if we know path parameters always occupy the beginning of `parameters`, then we can choose this strategy. Otherwise, let's adopt the more conservative one.